src/iterative/KOSMAPOSL/KOSMAPOSLReconstruction.cxx

```diff
@@ -0,0 +1,1162 @@
+                                / (2 * square(get_sigma_dp())));
+                      }
+                      else {
+                        kPET = 1;
+                      }
+
+                      kanatomical
+                        = exp(-(*this->kmnorm_sptr)[0][l][m]
+                              / square(anatomical_sd * sigma_m)
+                              / 2)
+                        * exp(-square(distance[dz][dy][dx] / grid_spacing.x())
+                              / (2 * square(sigma_dm)));
+
+                      kernelised_image_out[z][y][x]
+                        += kanatomical*kPET*image_to_kernelise[z+dz][y+dy][x+dx];
+
+                      pnkernel += kanatomical*kPET;
+                    }
+
+              if(current_alpha_estimate[z][y][x]==0){
+                continue;
+              }
+
+              kernelised_image_out[z][y][x]
+                = kernelised_image_out[z][y][x] / pnkernel;
+              pnkernel=0;
+
+
+            }
+        }
+    }
+}
+
+
+template<typename TargetT>
+void KOSMAPOSLReconstruction<TargetT>::compact_compute_kernelised_image(
+                         TargetT& kernelised_image_out,
+                         const TargetT& image_to_kernelise,
+                         const TargetT& current_alpha_estimate)
+{
+  // Something very weird happens here if I do not get_empty_copy()
+  // KImage elements will be all nan
+  unique_ptr<TargetT> kImage_uptr(current_alpha_estimate.get_empty_copy());
+  kernelised_image_out = *kImage_uptr;
+
+  const DiscretisedDensityOnCartesianGrid<3,float>* current_anatomical_cast =
+    dynamic_cast< const DiscretisedDensityOnCartesianGrid<3,float> *>
+      (this->get_anatomical_sptr ().get());
+  const CartesianCoordinate3D<float>& grid_spacing =
+    current_anatomical_cast->get_grid_spacing();
+
+  double kPET =0;
+  int min_dz, max_dz,min_dx,max_dx, min_dy,max_dy;
+
+  if (only_2D) {
+    min_dz = max_dz = 0;
+  }
+  else {
+    min_dz = -(num_neighbours-1)/2;
+    max_dz = (num_neighbours-1)/2;
+  }
+  min_dy = -(num_neighbours-1)/2;
+  max_dy = (num_neighbours-1)/2;
+  min_dx = -(num_neighbours-1)/2;
+  max_dx = (num_neighbours-1)/2;
+
+  Array<3,float> distance = Array<3,float>(IndexRange3D(min_dz, max_dz,
+                                                        min_dy, max_dy,
+                                                        min_dx, max_dx));
+
+  for (int z=min_dz;z<=max_dz;++z)
+    for (int y=min_dy;y<=max_dy;++y)
+      for (int x=min_dx;x<=max_dx;++x)
+        { // the distance is the euclidean distance:
+          //at the moment is used only for the definition of the neighbourhood
+          distance[z][y][x] =
+            sqrt(square(x*grid_spacing.x())+
+                 square(y*grid_spacing.y())+
+                 square(z*grid_spacing.z()));
+        }
+
+  // get anatomical standard deviation over all voxels
+
+  // calculate kernelised image
+
+  const int min_z = (*anatomical_sptr).get_min_index();
```

Comment:
Should `{min,max}_{x,y,z}` be calculated using `current_alpha_estimate` as in the full method?

Reply:
@danieldeidda

Reply:
it is actually the same since the two images have the same dimensions. actually, it might be better to write as private member min and max because it is reused in different places